### PR TITLE
refactor certificate controller to break it into two parts

### PR DESCRIPTION
**Pull request by mikedanese**
Break pkg/controller/certificates into:
* pkg/controller/certificates/approver: containing the group approver
* pkg/controller/certificates/signer: containing the local signer
* pkg/controller/certificates: containing shared infrastructure
```release-note
Break the 'certificatesigningrequests' controller into a 'csrapprover' controller and 'csrsigner' controller.
```

### CODE REVIEW

**Comment by luxas** [comment]
Fixes: https://github.com/kubernetes/kubernetes/issues/45302 I suppose?

Does this help https://github.com/kubernetes/kubernetes/issues/45030 as well?

**Comment by pipejakob** [comment]
Looks good so far. Will you continue the pattern from the GKESigner of adding events to the CSR on these different failure conditions? Would be very useful to trace/diagnose cluster issues.

Also, is the end goal to pull these out into completely different processes, or just separate goroutines in the controller-manager?

**Comment by mikedanese** [comment]
@pipejakob that's a good idea. I'm going to encorporate it in a follow up.

**Comment by mikedanese** [comment]
https://github.com/kubernetes/kubernetes/issues/46130

**Comment by pipejakob** [comment]
Cool. Are you looking for a stronger LGTM before restoring the tests, or are you set for now?

**Comment by mikedanese** [comment]
Not yet. Sorry.

**Comment by mikedanese** [comment]
Will ping explicitly.

**Comment by mikedanese** [comment]
@pipejakob this is ready for final review. I partially restored controller_test.go but a lot of what it is testing is duplicated in cfssl_signer_test.go, so I didn't reinclude that. PTAL.

**Comment by pipejakob** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by pipejakob** [comment]
Federation failure looks to be tracked in https://github.com/kubernetes/kubernetes/issues/45978.

File: pkg/controller/certificates/doc.go [hunk 1]
```diff
@@ -1,19 +0,0 @@
...
-
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
-
-// Package certificates contains logic for watching and synchronizing
-// CertificateSigningRequests.
```
Comment (pipejakob) [inline pkg/controller/certificates/doc.go]: With this file gone, and a lot of the code being moved into two new packages, they could each use their own package-level godoc (either inline in an existing file, or a `doc.go` file).

File: pkg/controller/certificates/approver/groupapprove_test.go [hunk 2]
```diff
@@ -14,54 +14,54 @@ See the License for the specific language governing permissions and
...
+package approver
 
 import (
 	"testing"
 
-	certificates "k8s.io/kubernetes/pkg/apis/certificates/v1beta1"
+	api "k8s.io/kubernetes/pkg/apis/certificates/v1beta1"
 )
 
 func TestHasKubeletUsages(t *testing.T) {
```
Comment (pipejakob) [inline pkg/controller/certificates/approver/groupapprove_test.go]: The only tests I can see here are for the `hasExactUsages()` helper, but not for the groupapprover code itself.

**Review by pipejakob — COMMENTED** [review]
Mostly LGTM, just a few minor comments.

File: pkg/controller/certificates/approver/groupapprove_test.go (hunk 2 above)
Comment (mikedanese) [inline pkg/controller/certificates/approver/groupapprove_test.go]: I added a bunch of test in https://github.com/kubernetes/kubernetes/pull/45619 for this.

File: pkg/controller/certificates/doc.go (hunk 1 above)
Comment (mikedanese) [inline pkg/controller/certificates/doc.go]: Added godoc to the packages.

**Comment by mikedanese** [comment]
@pipejakob federation is not a blocking build. I think I addressed all comments.

File: pkg/controller/certificates/approver/groupapprove_test.go (hunk 2 above)
Comment (pipejakob) [inline pkg/controller/certificates/approver/groupapprove_test.go]: Ah, I hadn't even seen https://github.com/kubernetes/kubernetes/pull/45619. I'll take a look.

**Comment by pipejakob** [comment]
Oh actually, @liggitt had a good comment in https://github.com/kubernetes/kubernetes/pull/45619. You renamed the controllers (and made the names shorter, as he recommended), but I don't see a release note about breaking backwards compatibility with the CLI.

**Comment by mikedanese** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by mikedanese** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by mikedanese** [comment]
@k8s-bot pull-kubernetes-unit test this

**Pull request merged**